### PR TITLE
GitHub Actions workflow for running Automated Tests

### PR DESCRIPTION
**Pull request by MukuFlash03**
The eventual goal is to have minimal manual intervention when testing latest images.
Essentially, we want to run the automated tests on every rebuild of the image so we know if the images are good enough to merge.

There are two scripts involved in running the demo tests: 
- [`demo-iso15118-2-ac-plus-ocpp.sh`](https://github.com/EVerest/everest-demo/blob/main/demo-iso15118-2-ac-plus-ocpp.sh)
- [`demo-automated-testing.sh`](https://github.com/EVerest/everest-demo/blob/main/demo-automated-testing.sh)

Another requirement is that we need to run tests on different operating systems, hence the GitHub actions workflow must run on atleast ubuntu (for now) and MacOS later on. 
MacOS with M1 chips has faced issues in the past and could be tricky to get configured.

------

Initial [PR](https://github.com/EVerest/everest-demo/pull/23) by shankari.
Manual testing steps highlighted in [this PR](https://github.com/EVerest/everest-demo/pull/58#issuecomment-2174585403).

### CODE REVIEW

**Comment by MukuFlash03** [comment]
### Related Commits / PRs
Listing some possibly related commits from the codebase history which could be helpful

--------

Demo Automated Tests: [PR link](https://github.com/US-JOET/everest-demo/pull/2)

------

“pip install pytest”: Apr 15, 2024 (1-3)
1. [Added this line](https://github.com/EVerest/everest-demo/pull/39/commits/2a5f4fd6b1356bdc0df66384d70154fe2da1361a)
2. [Removed this line](https://github.com/EVerest/everest-demo/pull/39/commits/8b63180a87d8305eaf231205f2765017027ad7e0)
3. [Added this line](https://github.com/EVerest/everest-demo/pull/39/commits/c7106eff70bbf30a0f8196b913e68190065c0d98)

------

tests/startup_tests.py: Dec 6, 2023 (1-2) ; Apr 15, 2024 (3)
1. Added: https://github.com/EVerest/everest-demo/commit/381a3cec855339cd6332cc751c0f1e9594101276
2. Modified: https://github.com/EVerest/everest-demo/commit/954b5d2dc44237e1bdf93eae3e5c1004fe268526
3. Removed: https://github.com/EVerest/everest-demo/pull/39/commits/8b63180a87d8305eaf231205f2765017027ad7e0

------

Scripts updated to use file path “ext/source/” : [commit link](https://github.com/EVerest/everest-demo/pull/39/commits/4fb6194267d178e42005e7aadb6dede9dfc797be)

------

Automated test docker compose file, bash script, run-test.sh: Dec 6, 2023 (1)
- [Added first here](https://github.com/EVerest/everest-demo/commit/a8ac948bc11eac354126318d83b05929477f674b)
- run-test.sh does not contain “$ pip install pytest” when it was first added.

------

Automated Tests related [PR link](https://github.com/EVerest/everest-demo/pull/5)

-------

Some more commits: 
- Removed macOS since docker not pre-installed: [commit link](https://github.com/US-JOET/everest-demo/commit/64872a5843a1f4d5c3bcc813a164190780a985cf)
- Switched to macOS completely; removed ubuntu: [commit link](https://github.com/US-JOET/everest-demo/commit/5d422599245eec793c78fcbfd0c55f54c85efa67)
- Host folder /var/folders added: [commit link](https://github.com/US-JOET/everest-demo/commit/304e16c27248da9a00c4d0815e863191a7661ce8)
- Timeout temporary, can be removed once automated tests added: [commit link](https://github.com/US-JOET/everest-demo/commit/f33a589c2eb0386f1cc5957e8754a4581fdaa256)
- Timeout bumped to 30 mins; enough time for 2nd .sh file to run: [commit link](https://github.com/EVerest/everest-demo/pull/23/commits/05528d4e76c1dc69f764bfb83459c88624613d3c)

**Comment by MukuFlash03** [comment]
### Starting off with getting demo-automated-testing.sh to run

----------

### Docker setup process in ubuntu vs in macOS
- Ubuntu: Don’t need to install docker as it looks like it comes pre-installed.
- macOS: Using a custom github action available [here](douglascamata/setup-docker-macos-action: A Github Action to setup Docker on macOS runners.).

Specific details available here in the [official list of runner images](https://github.com/actions/runner-images/tree/main).

-------

### Q. Where does startup_tests.py come from? It isn't there in this repo.

For the commit where startup_tests.py was first added, saw this comment:

```
# TODO: This should be removed once added to everest-core
COPY ./tests/startup_tests.py /ext/source/tests/core_tests/startup_tests.py
```

So, it looks like, the absence of the file should not cause any errors since it should be there in everest-core.
Yes, startup_tests.py present [here in everest-core](https://github.com/couryrr-afs/everest-core/blob/main/tests/core_tests/startup_tests.py)

Also, looking at [Dockerfile](https://github.com/EVerest/everest-demo/blob/main/manager/Dockerfile) in `everest-demo/manager/` , I see that it clones the `everest-core` git repo, copies `run-test.sh` to `/ext/source` which is a folder created in the Dockerfile that renames` everest-core/` to `/ext/source`.

**Comment by MukuFlash03** [comment]
### Timeout causing jobs to get cancelled

-------

So I found that timeout leads to jobs being cancelled:
- Timeout temporary, can be removed once automated tests added: https://github.com/US-JOET/everest-demo/commit/f33a589c2eb0386f1cc5957e8754a4581fdaa256

**Comment by MukuFlash03** [comment]
### Demo-automated-testing.sh passed for ubuntu

-----

Here is the successful: [workflow run](https://github.com/MukuFlash03/everest-demo/actions/runs/9605387730/job/26492908256)

This is output seen in the workflow logs:
```
manager-1      | ============================== 2 passed in 18.61s ==============================
```

I hope this is the right indicator that the current automated tests defined by the test script have passed.

**Comment by MukuFlash03** [comment]
### Added macOS runner to run automated tests

-------

MacOS docker setup passed finally using the same custom github action used in [this PR](https://github.com/EVerest/everest-demo/pull/23). 
I did have to change the macOS version from `macos-latest` to `macos-13` as the github action to setup docker wasn't compatible with the latest `macos-14`.

-----

Timeout increased to 60 minutes
- Increasing the timeout since the Downloading and Extracting steps themselves have taken about 20 minutes in the macOS runner.
- But right now even for the demo-automated-tests.sh on MacOS its taking close to 40-45 minutes just for the docker image download + extraction to complete. 
- This is just for the macOS runner; ubuntu job completes pretty quickly.

As seen in [this commit](https://github.com/EVerest/everest-demo/pull/23/commits/05528d4e76c1dc69f764bfb83459c88624613d3c) as well, when Shankari changed the timeout to 30 minutes, it was because the demo-iso15118-2-ac script itself was taking close to 30 minutes.

**Comment by MukuFlash03** [comment]
### Observations for Workflow Runs with macOS

-------

Test 1:

This macOS [run passed](https://github.com/MukuFlash03/everest-demo/actions/runs/9619753931/job/26536572676) - Pytest tests passed successfully.
However, the download and extraction of docker image itself takes close to 30 minutes, so had to increase timeout to 60 minutes.

-------

Test 2: 

1. Currently a PR is open into my local branch - automate-tests-merge; hence pushes trigger workflow for this as well.
[cicd.yaml run](https://github.com/MukuFlash03/everest-demo/actions/runs/9623533692).

2. But since this is a PR, the e2etest.yaml correctly skipped the jobs when triggered by a PR:
[e2etest.yaml run](https://github.com/MukuFlash03/everest-demo/actions/runs/9623535550).

3. Then added workflow dispatch to so that it triggers whenever cicd.yaml workflow is completed.
Ensuring that the dispatch job runs only when docker job completes successfully in cicd workflow by using “needs” attribute.
[cicd.yaml run](https://github.com/MukuFlash03/everest-demo/actions/runs/9623533451).

4. Successfully triggered via the workflow run that itself was triggered by a non-PR event (push commit).
[e2etest.yaml run](https://github.com/MukuFlash03/everest-demo/actions/runs/9623535552).

However in this macOS run, 1 test failed in the pytest startup_tests.
Usually there are two tests and both pass.

1 test failed with error message:
```
manager-1      |         if status == None or len(status) == 0:
manager-1      | >           raise TimeoutError("Timeout while waiting for EVerest to start")
manager-1      | E           TimeoutError: Timeout while waiting for EVerest to start
manager-1      | 
manager-1      | /usr/lib/python3.10/site-packages/everest/testing/core_utils/everest_core.py:198: TimeoutError
```

-------

Test 3:

Testing again by pushing changes.
Another different error observed in docker setup process for macOS in [this run](https://github.com/MukuFlash03/everest-demo/actions/runs/9624169333/job/26547511657).

-------

Test 4:

Testing again in [this run](https://github.com/MukuFlash03/everest-demo/actions/runs/9629869433/job/26559803309), specifically to observe MacOS runner.
Docker setup successful.

However, yet again 1 test failed, 1 test passed.
But the error message was different from the error seen in previous run:


```
manager-1      | >       assert probe.test(20)
manager-1      | E       assert False
manager-1      | E        +  where False = <bound method ProbeModule.test of <startup_tests.ProbeModule object at 0x7ebd56d50a90>>(20)
manager-1      | E        +    where <bound method ProbeModule.test of <startup_tests.ProbeModule object at 0x7ebd56d50a90>> = <startup_tests.ProbeModule object at 0x7ebd56d50a90>.test
manager-1      | 
manager-1      | core_tests/startup_tests.py:101: AssertionError
```

**Comment by MukuFlash03** [comment]
### Queries before moving onto demo-iso15118-2-ac-plus-ocpp.sh

**Which version of the bash script parameters to use?**
**Should all the command versions run in the workflow?**

-----

Shankari’s command for the AC script was valid as per [this Readme.md](https://github.com/US-JOET/everest-demo/tree/turn_on_basic_testing#step-1-run-the-demo). 

```
🚨 Basic and ISO 15118-2 AC Charging with OCPP 201 CSMS (MaEVe) ⚡: 
$ curl https://raw.githubusercontent.com/everest/everest-demo/main/demo-iso15118-2-ac-plus-ocpp201.sh | bash
```

----

But [latest code](https://github.com/EVerest/everest-demo#step-1-run-the-demo) in EVerest/everest-demo parent repo has the new script which was added in the single demo script added in [this commit](https://github.com/EVerest/everest-demo/commit/45ca2f3).

```
🚨 Basic and ISO 15118-2 AC Charging with OCPP 2.0.1 CSMS (MaEVe Security Profile 1) ⚡: 
$ curl https://raw.githubusercontent.com/everest/everest-demo/main/demo-iso15118-2-ac-plus-ocpp.sh | bash -s -- -1
```

Bash command has different arguments passed to choose which demo to run.

Should all the commands run in the workflow?
But these scripts run docker containers without -d flag which causes them to stay up and running.
Even if -d flag was supplied, they might be using the same port so would need to stop the containers before executing other scripts.

**Comment by MukuFlash03** [comment]
### Concerns with image build push workflow

-----

I get that the requirement is to run the automated tests on every rebuild of the images so that we know if the changes are good to run.
However, in the cicd.yaml file, I see that the latest images are only pushed to the GHCR registry on a non-pull-request event.

I’ve even looked at the actions history and can verify that for PRs, the push value is set to false in the docker-build-push action.

```
Run docker/build-push-action@v5
  with:
    context: ./manager
    push: false
    tags: [ghcr.io/everest/everest-demo/manager:0.0.16](http://ghcr.io/everest/everest-demo/manager:0.0.16)
  [ghcr.io/everest/everest-demo/manager:latest](http://ghcr.io/everest/everest-demo/manager:latest)
```

This results in images being built but not pushed to the GHCR as can be observed in the Build and Push step in the cicd.yaml workflow
```
WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
```

This is observed even for PRs that triggered workflow multiple times since multiple pushes were made:
- https://github.com/EVerest/everest-demo/actions/runs/9271474479/job/25506930245
- https://github.com/EVerest/everest-demo/actions/runs/9271482531

------

Now this is a concern, since the single line demo scripts make use of the docker-compose YAML files to run containers. 
These pulls images from the GHCR registry using the tag name available in .env file.
Now this tag name will still refer to the image without the latest changes from the PR.
This is because as specified above, the image will only be pushed if the trigger event is not a PR.

Hence, the demo scripts tests will actually run on the current latest tagged image but not on the latest code as it has no relevant tag right now.
From what I see the flow probably is that once a non-PR or normal push or commit is made to main branch, then the image is built and pushed to the GHCR registry and then the tags can be updated.
Now, next time onwards, the code will use this latest image.

So in current scenario, for the tests to run on the latest code, the image must be pushed once and tags updated before running e2etest.yaml workflow.

**Comment by shankari** [comment]
@MukuFlash03 

First, I don't want a giant portmanteau PR that is hard to review and merge. This PR should do one thing and do it well, which is enabling automated tests. The `demo-iso15118-2-ac-plus-ocpp.sh` script should be enabled, but in a separate PR.

High level comments:
>  Essentially, we want to run the automated tests on every rebuild of the image so we know if the images are good enough to merge.

This is only partially correct. We do want to test out new images, but not all changes are to the images. Some changes are to the demo scripts as well. Note that the PR that required multiple rounds of testing from @louisg1337 did not make any changes to the images.

> 1. Which version of the bash script parameters to use?

Obviously the most recent one. Shankari's PR was started before the script was refactored. It wasn't changed since then. We should use the currently valid version of the script. However, we are going to focus on only the automated testing script for now, so this is irrelevant for this PR.

> Found a [related commit here](https://github.com/US-JOET/everest-demo/commit/f33a589c2eb0386f1cc5957e8754a4581fdaa256) that highlights this issue with the main reason for using timeout being that the docker compose scripts run without the -d flag and hence the process stays active in the terminal.

The challenge with using `-d` is that then we won't see the logs - the container runs completely in the background. So if the test fails, we won't be able to see why. It also means that the action/job will not fail if the test fails. The only way to check whether the test actually passed or failed is to look at the logs, which is not how tests are supposed to work.
I would suggest using https://stackoverflow.com/questions/33799885/how-to-stop-all-containers-when-one-container-stops-with-docker-compose to ensure that the action sees the exit code from the test. Please also indicate the testing done, including a success and failure test case (you can change the entrypoint temporarily for this testing).

> 2. Should all the command versions run in the workflow?
> But these scripts run docker containers without -d flag which causes them to stay up and running.
Even if -d flag was supplied, they might be using the same port so would need to stop the containers before executing other scripts.

I am not sure what you mean by this. each separate run in the matrix runs in a separate VM. So I am not sure what you mean by the 'same port"

> I get that the requirement is to run the automated tests on every rebuild of the images so that we know if the changes are good to run.

As I said above, this is necessary but not sufficient

> However, in the cicd.yaml file, I see that the latest images are only pushed to the GHCR registry on a non-pull-request event.

This is expected, since we only want to push valid images to the repo.

>  So in the current scenario, for the tests to run on the latest code, the image must be pushed once and tags updated before running e2etest.yaml workflow.

That is not the only option. A better option is to test the locally built image
1. We don't need to rebuild the image if the only change is to the demo files. In this case, it is fine to run the demo script directly.
2. If there are changes to the non-demo files, we do need to rebuild the images. And we do. We just don't push them. We should be able to test against the locally built images from the previous jobs. Please read up on docker images pull and cache - similar to git repos, if you don't explicitly specify pull, docker will only pull the remote image if it does not exist locally. We have just built these images as part of the github action, so when you run the tests, if you do so in the same environment, they will use the newly built versions.

File: .github/workflows/cicd.yaml [hunk 1]
```diff
@@ -32,74 +31,95 @@ jobs:
...
+      - name: Temporary step to pass the job
+        id: temporary-step
+        run: echo "Temporary step to ensure job completes and workflow doesn't fail"
+
+      # - name: Checkout
+      #   uses: actions/checkout@v4
+      #   with:
+      #     fetch-depth: 0
+
+      # - name: Ensure Docker image version is not referencing an existing release
```
Comment (shankari) [inline .github/workflows/cicd.yaml:43]: why is all this commented out?

**Review by shankari — COMMENTED** [review]
I am not quite sure what these changes are doing or why the image build and push has been commented out.

**Comment by shankari** [comment]
Finally,

I am not sure why this change was introduced:

> Removed macOS since docker not pre-installed: [commit link](https://github.com/US-JOET/everest-demo/commit/64872a5843a1f4d5c3bcc813a164190780a985cf)
> Switched to macOS completely; removed ubuntu: [commit link](https://github.com/US-JOET/everest-demo/commit/5d422599245eec793c78fcbfd0c55f54c85efa67)

when the goal is 

> Another requirement is that we need to run tests on different operating systems, hence the GitHub actions workflow must run on atleast ubuntu (for now) and MacOS later on.

Note also that, given the time and resources required to run MacOS, I don't think we should run it on every pull request. The idea behind containerization/software based testing is that if the code works in one container, it should work in all containers. This is clearly not true 100% of the time, but if it works in one setup, we know that the application logic is right, and we just have to figure out the system-level library incompatibilities.

**Comment by MukuFlash03** [comment]
> Finally,
> 
> I am not sure why this change was introduced:
> 
> > Removed macOS since docker not pre-installed: [commit link](https://github.com/US-JOET/everest-demo/commit/64872a5843a1f4d5c3bcc813a164190780a985cf)
> > Switched to macOS completely; removed ubuntu: [commit link](https://github.com/US-JOET/everest-demo/commit/5d422599245eec793c78fcbfd0c55f54c85efa67)
> 
> when the goal is
> 
> > Another requirement is that we need to run tests on different operating systems, hence the GitHub actions workflow must run on atleast ubuntu (for now) and MacOS later on.

Just to clarify, these are not changes that I introduced. These were commits that I found in [this PR](https://github.com/EVerest/everest-demo/pull/23) by Shankari and was just listing the relevant commits that could help better understand the task at hand.  


> Note also that, given the time and resources required to run MacOS, I don't think we should run it on every pull request. The idea behind containerization/software based testing is that if the code works in one container, it should work in all containers. This is clearly not true 100% of the time, but if it works in one setup, we know that the application logic is right, and we just have to figure out the system-level library incompatibilities.

Alright, noted.

**Comment by MukuFlash03** [comment]
Based on the previous conversations, I tried a few approaches and listing them down along with reasons why I didn't use a specific approach and finally stuck with one.
One of the primary tasks was to build the locally and then test it.

Removed e2etest.yaml file as it didn’t seem relevant to run a separate workflow after cicd.yaml.

-----

Approach 1:
- Added a 2nd job to run automated tests after 1st job that builds and pushes docker images in cicd.yaml.
- Now while this seemed good initially, I found some concerns with these:
- - In case of a PR event, it’s okay to have the build + push first, then test.
- - This is because since push isn’t executed, testing will run on the built image.

- - But in case of non-PR or push event when the PR is merged, the push condition evaluates to true and if test is done after pushing image, the test may fail.
- - And the erroneous image would have already been pushed to the registry as a part of the prior build + push action step.

Hence not moving ahead with this approach.

-------

Approach 2:
- Revised workflow jobs to follow this: Build -> Test -> Build + Push (existing)
- The first build job would build image with the push attribute set to false as we don’t want to push even in case of a non-PR event before we test the image.
- The image is then tested in the 2nd job that runs the automated tests.
- Then proceed to Build + Push that uses the cache.

The problem with this was that the docker compose was unable to retrieve the cached image layers that were built in the first Build job.
So adding just another build job with duplicated code from the 3rd job for Build + Push wasn’t useful.

-----

Approach 3:
- Revised workflow: Build + Test -> Build + Push
- Now, the 1st job is actually the run-automated-tests job that uses the bash script to build the image locally using the docker compose commands.
- Had to add the build attribute to the docker compose services.
- Also added abort-on-exit and exit-code flags that help in killing the running docker compose up command once any container exits.
- Additionally, if non-zero exit code is returned, the entire workflow fails before 2nd job that builds and pushes image.
- Hence we can be assured that latest image is tested first before being pushed.

Finally decided to go ahead with Approach 3.

**Comment by MukuFlash03** [comment]
### Testing Done in my Forked Repo 

Done by Triggering Workflow Runs for a Success and Failure Test Case
 
> I would suggest using https://stackoverflow.com/questions/33799885/how-to-stop-all-containers-when-one-container-stops-with-docker-compose to ensure that the action sees the exit code from the test. Please also indicate the testing done, including a success and failure test case (you can change the entrypoint temporarily for this testing).

------

Temporarily changed entrypoint in [docker-compose.automated-tests.yml](https://github.com/MukuFlash03/everest-demo/commit/1b0e1f2fb9cdc3cba211ff785f485d05da0daf0b#diff-44f88d2989071c4f7f8be705b3485e0f627dbd79e9b9dc5b4bc0907044e9a61c) for the purposes of this testing

```
- entrypoint: "sh ./run-test.sh"
+ entrypoint: "sh ./test-exit-code.sh success" 
```

Similarly triggered a separate run for failure:
```
+ entrypoint: "sh ./test-exit-code.sh failure" 
```
-------

[Test script](https://github.com/MukuFlash03/everest-demo/commit/1b0e1f2fb9cdc3cba211ff785f485d05da0daf0b#diff-40f18b4d1f4d2d9821a517751fd7eb6d787e433dc0504d9baefc278adffe64a5) used for testing
```
#!/bin/bash
TEST_TYPE=$1
exit_code=0
if [ "$TEST_TYPE" == "success" ]; then
    exit_code=0
elif [ "$TEST_TYPE" == "failure" ]; then
    exit_code=1
fi

echo "Exit Code from test-exit-code.sh: $exit_code"
exit $exit_code
```

------

Success Test Case:
- [commit](https://github.com/MukuFlash03/everest-demo/commit/1b0e1f2fb9cdc3cba211ff785f485d05da0daf0b): passing `success` as a parameter to the entrypoint script.
- [workflow run](https://github.com/MukuFlash03/everest-demo/actions/runs/9784262516/job/27014762129#step:4:4037)
- workflow continues successfully after the run-automated-tests job and moves onto complete the following jobs.

-----

Failure Test Case 
- [commit](https://github.com/MukuFlash03/everest-demo/commit/7623345ee86ad651921dd1be0a651edb5a5502b3): passing `failure` as a parameter to the entrypoint script.
- [workflow run](https://github.com/MukuFlash03/everest-demo/actions/runs/9784315072/job/27014924098#step:4:4071)
- workflow fails immediately as soon as non-zero exit code is received in the `manager` service and skips all jobs.

**Comment by shankari** [comment]
> In the demo-automated-testing.sh script ([commit](https://github.com/EVerest/everest-demo/pull/62/commits/36d7c1afbfaa759b77fde744dba14562e922cfc1#diff-555f9d0cb98b05327f874a5199c6929acdfaf1ad5e5af62a20cf592d0afff6c9L34-R35)), I had to change the docker compose command to use the docker-compose.automated-tests.yml directly via $DEMO_COMPOSE_FILE_NAME instead of $DEMO_DIR/DEMO_COMPOSE_FILE_NAME.

@MukuFlash03 This is not correct. I am not sure why you are running the `demo-XXX` script in the first place. The demo scripts are designed to be "single line demos" that people can run without having to check out any code. They essentially check out the repo in a separate temporary directory and then run docker. But in the case of GitHub actions, we have already checked out the repo. It is not clear why you would run the demo script (which is a lightweight wrapper) and then edit it to remove its original function instead of just running docker compose.

File: .github/workflows/cicd.yaml [hunk 2]
```diff
@@ -2,16 +2,47 @@ name: cicd
 
 on:
   pull_request:
-    branches:
-      - main
+    branches: [ main ]
```
Comment (shankari) [inline .github/workflows/cicd.yaml:5]: extraneous change, can be reverted

File: .github/workflows/cicd.yaml [hunk 3]
```diff
@@ -2,16 +2,47 @@ name: cicd
...
   push:
-    branches:
-      - main
+    branches: [ main ]
```
Comment (shankari) [inline .github/workflows/cicd.yaml:7]: ditto

File: demo-scripts/charging-profile-0.json
Comment (shankari) [inline demo-scripts/charging-profile-0.json:1]: why do we need this file?

File: demo-scripts/charging-profile-1.json
Comment (shankari) [inline demo-scripts/charging-profile-1.json:1]: or this one?

File: demo-scripts/maeve-set-charging-profile.sh
Comment (shankari) [inline demo-scripts/maeve-set-charging-profile.sh:1]: or this one. All of these seem to be from a separate change by @louisg1337

File: .github/workflows/cicd.yaml [hunk 4]
```diff
@@ -2,16 +2,47 @@ name: cicd
...
+      - name: List Docker images
+        run: docker images
+
+      - name: Run automated tests using docker-compose.automated-tests.yml
+        timeout-minutes: 30
+        run: |
+          echo "Running docker compose up..."
+          docker compose --project-name everest-ac-demo \
+                  --file "docker-compose.automated-tests.yml" up \
+                    --build \
```
Comment (shankari) [inline .github/workflows/cicd.yaml:30]: Why are we building this separately? We build the image later (https://github.com/EVerest/everest-demo/pull/62/files#diff-2c9df400afaffa9270cab617792f08cc671e5f95ae192f1cf111ef4e990a4addL97) and the build takes *a long time*. We should not rebuild unless we have to.

I see that you wrote a fairly long description of the options here
https://github.com/EVerest/everest-demo/pull/62#issuecomment-2206803270

but I wonder if you checked the documentation for the build-push action, and the "test before push" in particular
https://github.com/docker/build-push-action?tab=readme-ov-file

File: .github/workflows/cicd.yaml [hunk 5]
```diff
@@ -2,16 +2,47 @@ name: cicd
...
+                    --abort-on-container-exit \
+                    --exit-code-from manager
+
+          exit_code=$?            
+          echo "Docker-compose up exit code from manager service: $exit_code"       
+
+          echo "Running docker compose down..."
+          docker compose --project-name everest-ac-demo \
+                  --file "docker-compose.automated-tests.yml" down
```
Comment (shankari) [inline .github/workflows/cicd.yaml:39]: I don't see how this action will fail if the test fails. we will just continue and build + push anyway. Have you actually tested that the workflow (not just the docker compose) will fail if the tests fail?

File: docker-compose.automated-tests.yml
Comment (shankari) [inline docker-compose.automated-tests.yml:1]: you should not need any changes to this file.

**Review by shankari — CHANGES_REQUESTED** [review]

File: .github/workflows/cicd.yaml (hunk 5 above)
Comment (shankari) [reply .github/workflows/cicd.yaml:39]: I see that you linked to a workflow where the workflow failed, but I don't understand how it works. The exit code of a step is the exit code of the last statement by default. If the manager service ends and then you run another docker compose after it, the exit code should be the exit code of the `docker compose down`. I would appreciate an explanation of why.

File: demo-scripts/maeve-set-charging-profile.sh
Comment (MukuFlash03) [reply demo-scripts/maeve-set-charging-profile.sh:1]: I'm honestly not sure how this happened, my PR was created after @louisg1337 's PR's merged, so the changes should have been there in my branch code already and not appear as commits. 

On comparing the timestamps, this is the [workflow run](https://github.com/EVerest/everest-demo/actions/runs/9782065706) that was triggered at the same timestamp now applied to all my commits.

The only possibility I can think of is I remember missing to sign-off some commits, and then ran the rebase command as per the documentation. Perhaps that somehow messed it up. 

I'll exclude the changes from this PR.

-----

Commits History

What's weird is that all my commits now have have the same timestamp of Jul 3rd, 10:08 AM.
I definitely did not make all the code changes and commits in a single day. 


Screenshot 2024-07-11 at 10 33 38 PM

----

Workflow Runs

On looking at the workflow runs, you can see they're spread out over 2-3 weeks ago. 

Screenshot 2024-07-11 at 10 33 23 PM

------

File: .github/workflows/cicd.yaml (hunk 5 above)
Comment (MukuFlash03) [reply .github/workflows/cicd.yaml:39]: Yes, you're correct about not needing `docker compose down`. 
I might have added it when testing the updated docker compose up with the different abort and exit code flags and forgot to remove it.

Also, yes, the exit code returns the last run command. 

-------

With regards to the working of the exit code mechanism:

- Initially, when I had incorrectly modified the `demo-automated-testing.sh` script, I had included the docker compose commands inside it. I needed to pass the exit code from the `docker compose up` command to GitHub actions and hence used a variable `exit_code=$?` to do that. 
- - Read up about `$?` in this [article](https://how.wtf/the-right-way-to-test-exit-codes-in-bash.html) and passing exit code via scripts [here](https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions).

- Now, without the demo-automated-testing.sh, when I'm directly executing the `docker compose up` inside the workflow file as a job step, it's still the same behavior.
-  - In the workflow runner instance, the `exit_code` variable still stores the exit code from the last executed command above it which is the `docker compose up` command. 

----

Now, bringing GitHub Actions into the picture and how it determines the exit_code to look at.

Setting exit code flag for shell : `set -e or +e flag` ([documentation](https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html)
- The `-e` flag instructs the shell to exit immediately if a non-zero exit code is encountered.
- The `+e` flag on the other hand, allows execution to continue despite receiving non-zero exit codes.
- In our case, we want the `-e` flag to be used.

**GitHub action runners** use the `-e` flag **by default** (mentioned in accepted answer in [this discussion](https://github.com/orgs/community/discussions/46992#discussioncomment-4961541))

-----

To summarize: 

So considering the two scenarios: 
- Success (exit code 0): Workflow run continues executing onto the steps after `docker compose up`.
- Failure (non-zero exit code): Workflow run fails immediately and successive steps, jobs not executed at all.

**Comment by MukuFlash03** [comment]
Addressing both these comments as they are related to the image build process:

1. Why make changes to `docker-compose-automated-tests.yml` ([link](https://github.com/EVerest/everest-demo/pull/62#discussion_r1674813302))
2. Why rebuilding? ([link](https://github.com/EVerest/everest-demo/pull/62#discussion_r1674812168))

The main reason I need changes to this file is because the github actions flow I'm proposing involves two jobs:
- Run automated tests: Uses docker compose up to Build and Run Tests.
- Build Push docker image.

-----

So, I needed to build the image locally first in the `docker compose up`.
This is enforced by using the build flag (`--build`).

I tried two scenarios where I had used only build flag ((in docker compose up command) or only build attribute (docker compose file). 

Workflow Run and Commit with only build flag in docker compose up command:
[Run](https://github.com/MukuFlash03/everest-demo/actions/runs/9912983164/job/27388979064): 
[Commit](https://github.com/MukuFlash03/everest-demo/blob/1a7121afeb418b19c9a4cfd9c6b2d795a9faf9a9/docker-compose.automated-tests.yml)

Workflow Run and Commit with only build attribute in docker compose file: 
[Run](https://github.com/MukuFlash03/everest-demo/actions/runs/9914002240/job/27392203170): 
[Commit](https://github.com/MukuFlash03/everest-demo/commit/04dff674598b0650a465e9a2bd901bd4623b8b05).

In both cases, I observed that the image is always pulled first and not built locally.

**Comment by MukuFlash03** [comment]
What I understood from this is:

Both `build flag` (in docker compose up command) and `build attribute` (docker compose file) are required to build image locally.

The reason for this is that:
- Even after using `--build`, the build context for the Dockerfile path must be provided.
- Also, the hierarchy that is followed when deciding whether to pull or build images locally, specifically when both `image` and `build` attributes are specified in a docker compose file.
This is mentioned in the [documentation](https://docs.docker.com/compose/compose-file/build/#using-build-and-image):

```
When Compose is confronted with both a build subsection for a service and an image attribute, it follows the rules defined by the pull_policy attribute.
If pull_policy is missing from the service definition, Compose attempts to pull the image first and then builds from source if the image isn't found in the registry or platform cache.
```

The priority of image source as per this above info is:
1. Local image is used directly.
2. If the image isn't found in the registry or platform cache.
	A. Compose attempts to pull the image first
	B. Then (re)builds from source.

But on observing the workflow runs, I saw that:
- This hierarchy not being followed.
- Image is always being built in our case with my changes.

The answer is the build flag in the docker compose up command.
This is forcing the image to be rebuilt every time.
I did test removing it and it pulled the image instead then.
But this is what we want as it the first time in the workflow that the images are being built.

So we do need the build flag and also build attribute to specify context containing Dockerfile.
Because we want to be able to test the latest code changes which would be present in the locally built image.

File: docker-compose.automated-tests.yml
Comment (MukuFlash03) [reply docker-compose.automated-tests.yml:1]: Addressing in [below comments](https://github.com/EVerest/everest-demo/pull/62#issuecomment-2226692349).

File: .github/workflows/cicd.yaml (hunk 4 above)
Comment (MukuFlash03) [reply .github/workflows/cicd.yaml:30]: Addressing in [below comments](https://github.com/EVerest/everest-demo/pull/62#issuecomment-2226692349).

**Comment by MukuFlash03** [comment]
2. Why rebuilding? (addressing [this comment](https://github.com/EVerest/everest-demo/pull/62#discussion_r1674812168))

**Comment by MukuFlash03** [comment]
So yes, while the workflow does work, tests run, containers exit....there is the main problem of the long build times.

Currently, both jobs build images. 
Where I am stuck is in finding a way to be able to do this:
```
Support Build and caching from docker-compose files
```

The advantage of this would be that the image layers built in the 1st job will be cached and these should be available to the 2nd job when it tries to build the image, thereby making the build process faster by not building in both jobs.

Found some documentation and articles on how this might be possible.
- Caching from docker-compose files ([link](https://github.com/docker/build-push-action/issues/128))
- Build Push action + Docker compose ([link](https://github.com/docker/build-push-action/issues/150)) 
- Artifacts ([link](https://github.com/docker/build-push-action/issues/147#issuecomment-703840959)): not good for large images, already discussed above.
- Official Docker Bake Action for GitHub ([link](https://github.com/docker/bake-action)) 


Taking a look at these next.

**Comment by shankari** [comment]
@MukuFlash03 I am not sure why you need all these links. we use docker build and push. docker build and push has support for testing in the middle. I linked to the documentation on testing. I would suggest you just use it and move on

**Comment by MukuFlash03** [comment]
> @MukuFlash03 I am not sure why you need all these links. we use docker build and push. docker build and push has support for testing in the middle. I linked to the documentation on testing. I would suggest you just use it and move on

I've gone ahead and made the change to implement the test before push workflow inside the docker build push job itself.

I did take a look initially and one reason I was trying an alternative method was the fact that we are dealing with matrix jobs.

More details in this comment.

**Comment by MukuFlash03** [comment]
The changes from this PR are now moved over to a new PR #67 
Closing this PR.

This was done to handle the commit history being messed up by the rebase commands to add signoffs to commits.
The issue is described in [this comment](https://github.com/EVerest/everest-demo/pull/62#discussion_r1675344451).
